### PR TITLE
Update Web Terminal Operator uninstall docs

### PR DESCRIPTION
**Pull request by amisevsk**
Update the Web Terminal Operator uninstall process doc to clarify a few things
* Update the list of CRDs to remove from the cluster to include all CRDs used by the Operator
* Clarify that depending on the installation, some optional resources may not be present (and the failure to delete them is not an issue)
* Verify and highlight removal of CRDs, as we've found that failing to fully remove CRDs before uninstalling the Operator can cause cluster-wide issues due to the use of conversion webhooks: https://issues.redhat.com/browse/OLM-2336

### CODE REVIEW

**Comment by amisevsk** [comment]
cc: @JPinkney

**Review by JPinkney — APPROVED** [review]

**Comment by amisevsk** [comment]
Hi @Preeticp, I bothered in my last PR here as well -- would it be possible to arrange review of this PR?

File: modules/odc-uninstalling-web-terminal.adoc [hunk 1]
```diff
@@ -72,7 +72,7 @@ $ oc delete devworkspaceroutings.controller.devfile.io --all-namespaces --all --
 If this step is not complete, finalizers make it difficult to fully uninstall the Operator easily.
 ====
 +
-. Run the following commands to remove the CRDs:
+. Run the following commands to remove the CRDs used by the operator:
```
Comment (sounix000) [inline modules/odc-uninstalling-web-terminal.adoc:75]: **Suggestion**: Start with the objective, instead of the action.

To remove the CRDs used by the Operator, run the following commands:

File: modules/odc-uninstalling-web-terminal.adoc [hunk 2]
```diff
@@ -84,6 +84,28 @@ $ oc delete customresourcedefinitions.apiextensions.k8s.io devworkspaceroutings.
...
+----
+$ oc delete customresourcedefinitions.apiextensions.k8s.io devworkspaceoperatorconfigs.controller.devfile.io
+----
++
+[WARNING]
+====
+The DevWorkspace Operator ships custom resource definitions that make use of conversion webhooks. Failing to remove the CRDs above can cause issues on the cluster, as the steps below will remove the server that handles these conversion webhooks.
+====
++
+. Verify that all involved custom resource definitions are removed. The command below should return no results.
```
Comment (sounix000) [inline modules/odc-uninstalling-web-terminal.adoc:102]: **Suggestion**: "The command below... " --> "Note that on successful removal, the following command should not display any result."

File: modules/odc-uninstalling-web-terminal.adoc [hunk 3]
```diff
@@ -108,9 +130,14 @@ If you remove the `devworkspace-webhook-server` deployment without removing the
 +
 . Run the following commands to remove any lingering services, secrets, and config maps:
```
Comment (sounix000) [inline modules/odc-uninstalling-web-terminal.adoc:131]: I wouldn't feel comfortable using the word "lingering". I would use "remaining", if that is technically accurate.

File: modules/odc-uninstalling-web-terminal.adoc [hunk 4]
```diff
@@ -108,9 +130,14 @@ If you remove the `devworkspace-webhook-server` deployment without removing the
...
 +
+[NOTE]
+====
+Depending on the installation, some of the resources below may not be present on the cluster.
```
Comment (sounix000) [inline modules/odc-uninstalling-web-terminal.adoc:135]: - Extra whitespace before "some".
- "... some of the resources below ..." ---> "... some resources included in the following command might not be available on the cluster."
- Instead of a formal note, I would use an ordinary sentence.

**Review by sounix000 — APPROVED** [review]
@amisevsk - I have pointed out some nitpicks. Overall, it looks good. Also, many thanks for contributing to documentation - we really appreciate it! I am pre-approving the changes requested, because I am sure you will make necessary changes (at least the ones that makes sense technically).

@rishumehra - Please feel free to make any suggestion, because this is largely a part of ODC docs.

File: modules/odc-uninstalling-web-terminal.adoc (hunk 1 above)
Comment (rishumehra) [reply modules/odc-uninstalling-web-terminal.adoc:75]: @sounix000 Just to clarify, do we need to write "run the following commands" as I think it is understood to the developer/admin that the following are commands only.

File: modules/odc-uninstalling-web-terminal.adoc (hunk 3 above)
Comment (rishumehra) [reply modules/odc-uninstalling-web-terminal.adoc:131]: @sounix000 +1

File: modules/odc-uninstalling-web-terminal.adoc (hunk 2 above)
Comment (rishumehra) [reply modules/odc-uninstalling-web-terminal.adoc:102]: @sounix000 **Suggestion** that 'Note' as a word should be avoided. Can we retain 'Verify' at the start and the rest of the sentence should change as per your suggestion? In my opinion, "Note" should be used only when we're writing an actual "Note".

File: modules/odc-uninstalling-web-terminal.adoc [hunk 5]
```diff
@@ -84,6 +84,28 @@ $ oc delete customresourcedefinitions.apiextensions.k8s.io devworkspaceroutings.
...
+----
++
+[source,terminal]
+----
+$ oc delete customresourcedefinitions.apiextensions.k8s.io devworkspaceoperatorconfigs.controller.devfile.io
+----
++
+[WARNING]
+====
+The DevWorkspace Operator ships custom resource definitions that make use of conversion webhooks. Failing to remove the CRDs above can cause issues on the cluster, as the steps below will remove the server that handles these conversion webhooks.
```
Comment (rishumehra) [inline modules/odc-uninstalling-web-terminal.adoc:99]: **Rephrase suggestion:** The DevWorkspace Operator transmits custom resource definitions (CRDs) that use conversion webhooks. Failing to remove the CRDs can cause issues on the cluster. The following steps will remove the server that handles these conversion webhooks.

File: modules/odc-uninstalling-web-terminal.adoc (hunk 4 above)
Comment (rishumehra) [reply modules/odc-uninstalling-web-terminal.adoc:135]: @sounix000 +1

File: modules/odc-uninstalling-web-terminal.adoc (hunk 5 above)
Comment (Preeticp) [reply modules/odc-uninstalling-web-terminal.adoc:99]: I would suggest `provides` instead of `transmits`. Rest of the suggested change looks good.

File: modules/odc-uninstalling-web-terminal.adoc (hunk 1 above)
Comment (sounix000) [reply modules/odc-uninstalling-web-terminal.adoc:75]: @rishumehra - You are correct. In that case, we can simply write: "Remove the CRDs used by the Operator.".

This suggestion can be implemented across the entire PR/module.

File: modules/odc-uninstalling-web-terminal.adoc (hunk 2 above)
Comment (amisevsk) [reply modules/odc-uninstalling-web-terminal.adoc:102]: Changed to 
> Verify that all involved custom resource definitions are removed. The following command should not display any result.

I think it's useful to mention the first part (custom resource definitions) to explain what is being checked.

File: modules/odc-uninstalling-web-terminal.adoc (hunk 4 above)
Comment (amisevsk) [reply modules/odc-uninstalling-web-terminal.adoc:135]: For the first point, by extra whitespace, do you mean it should be
> `Depending on the installation,  some of the resources`
?

Second point -- changed but I used `may not exist` instead of `not be available`. Available reads somewhat like "services may be unusable", etc.

**Comment by amisevsk** [comment]
@Preeticp applied suggestions as above, with a few minor differences (I hope I didn't miss any). Thank you for the review!

File: modules/odc-uninstalling-web-terminal.adoc [hunk 6]
```diff
@@ -84,6 +84,28 @@ $ oc delete customresourcedefinitions.apiextensions.k8s.io devworkspaceroutings.
...
++
+[source,terminal]
+----
+$ oc delete customresourcedefinitions.apiextensions.k8s.io devworkspaceoperatorconfigs.controller.devfile.io
+----
++
+[WARNING]
+====
+The DevWorkspace Operator provides custom resource definitions (CRDs) that use conversion webhooks. Failing to remove the CRDs above can cause issues on the cluster. The following steps will remove the server that handles these conversion webhooks.
+====
```
Comment (Preeticp) [inline modules/odc-uninstalling-web-terminal.adoc:100]: Two suggestions for this note:
1. "The DevWorkspace Operator provides custom resource definitions (CRDs) that use conversion webhooks. Failing to remove the CRDs can cause issues on the cluster." This part of the note should go as a warning right under the step "Remove the CRDs used by the Operator:"
Pls. note that I have removed the word 'above' from the warning and made the Operator title case in the step.
2. This step should be modified from: "Remove the `DevWorkspace-Webhook-Server` deployment, mutating, and validating webhooks:" to "Remove the `DevWorkspace-Webhook-Server` deployment, mutating, and validating webhooks that handle the conversion webhooks." That will take care of the last sentence in the warning.

File: modules/odc-uninstalling-web-terminal.adoc [hunk 7]
```diff
@@ -106,11 +128,11 @@ $ oc delete validatingwebhookconfigurations controller.devfile.io
 If you remove the `devworkspace-webhook-server` deployment without removing the mutating and validating webhooks, you will not be able use `oc exec` commands to run commands in a container on the cluster. After you remove the webhooks you will be able to use the `oc exec` commands again.
```
Comment (Preeticp) [inline modules/odc-uninstalling-web-terminal.adoc:128]: Not part of this PR but would be great if you could add the missing word 'to' here: not be able to use the

**Review by Preeticp — COMMENTED** [review]
Thank you @sounix000 and @rishumehra. @amisevsk a couple of nits and a suggestion, PTAL.
Otherwise looks, great!

File: modules/odc-uninstalling-web-terminal.adoc (hunk 4 above)
Comment (Preeticp) [reply modules/odc-uninstalling-web-terminal.adoc]: For the first point, I think he wants you to remove the extra space between "installation, some" if there is any.
'may not exist' sounds good, but I guess this change has not been made, as I see - 'might not be present on the cluster' in the PR.

**Comment by Preeticp** [comment]
@amisevsk please let us know which versions of OpenShift do these changes apply to?
Also, WTO continues to be TP right? As we have TP note for this feature.

File: modules/odc-uninstalling-web-terminal.adoc (hunk 6 above)
Comment (amisevsk) [reply modules/odc-uninstalling-web-terminal.adoc:100]: These are OpenShift object names, which are case-sensitive. I think it may be less clear to use proper title case here when the object on the cluster is named `devworkspace-webhook-server`.

File: modules/odc-uninstalling-web-terminal.adoc [hunk 8]
```diff
@@ -55,7 +55,7 @@ The DevWorkspace Operator functions as a standalone Operator and may be required
 ====
 
 .Procedure
-. Run the following commands to ensure that all `DevWorkspace` CRs are removed along with their related Kubernetes objects, such as deployments.
+. Remove `DevWorkspace` custom resources used by the operator, along with all related Kubernetes objects, such as deployments.
```
[the remark this answers is not in the document]
Comment (amisevsk) [reply modules/odc-uninstalling-web-terminal.adoc:58]: There are two contexts in which the `DevWorkspace` phrase is used -- in one, we are referring to the operator itself, which is called `DevWorkspace Operator`. In the other, we are referring to the names of kubernetes objects, which derive from the operator name but are in lower-snake-case (e.g. `devworkspace-webhook-server`). 

I don't think it's a good idea to use title case for the latter examples, as trying to `oc get` that resource will fail, e.g. 
```
$ oc delete deployment/DevWorkspace-Webhook-Server -n openshift-operators
```
will return 
> Error from server (NotFound): deployments.apps "Devworkspace-Webhook-Server" not found

even if the deployment `devworkspace-webhook-server` is exists. Given the note we have that some resources may not exist on the cluster, this looks like a recipe for trouble.

File: modules/odc-uninstalling-web-terminal.adoc (hunk 6 above)
Comment (amisevsk) [reply modules/odc-uninstalling-web-terminal.adoc:100]: I've moved the warning up as suggested and dropped the piece about conversion webhooks, as this is technically no longer accurate. It's the main controller deployment that services these webhooks, and it is removed when the operator is uninstalled.

**Comment by amisevsk** [comment]
>please let us know which versions of OpenShift do these changes apply to?

These changes should be applied to OpenShift 4.8 and above. The process is also still valid for OpenShift 4.7, and can be applied there, as one added detail (the warnings about CRDs) is useful to have.

>Also, WTO continues to be TP right? As we have TP note for this feature.

At the moment, yes, WTO is still in TP. We hope to move to GA for OpenShift 4.10

File: modules/odc-uninstalling-web-terminal.adoc (hunk 7 above)
Comment (amisevsk) [reply modules/odc-uninstalling-web-terminal.adoc]: Good catch! Added.

File: modules/odc-uninstalling-web-terminal.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (rishumehra) [reply modules/odc-uninstalling-web-terminal.adoc:58]: @amisevsk Throughout our docs we use 'Operator' in the title case only in sentences/content only. Whereas in code/commands blocks you can use 'Operator' in any case because that time is being used as an example.

[the remark this answers is not in the document]
Comment (Preeticp) [reply modules/odc-uninstalling-web-terminal.adoc:58]: Hey @amisevsk  I think you misunderstood. 
I meant add the article 'the' before the word `DevWorkspace`
And wrt the case, in the word Operator, O needs to be capital throughout the PR. Right now in some places the pr has 'operator' we need to make it 'Operator'

File: modules/odc-uninstalling-web-terminal.adoc (hunk 6 above)
Comment (Preeticp) [reply modules/odc-uninstalling-web-terminal.adoc]: @amisevsk I agree. We need to change the case of the word 'operator' to 'Operator' when it is part of the general text as per OCP convention. If it is part of the name of a CRD, or part of a command, we need to retain it as is.

File: modules/odc-uninstalling-web-terminal.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (Preeticp) [reply modules/odc-uninstalling-web-terminal.adoc:58]:
```suggestion
. Remove the `DevWorkspace` custom resources used by the Operator, along with all the related Kubernetes objects, such as deployments.
```

File: modules/odc-uninstalling-web-terminal.adoc [hunk 9]
```diff
@@ -103,14 +125,14 @@ $ oc delete validatingwebhookconfigurations controller.devfile.io
 +
 [NOTE]
 ====
-If you remove the `devworkspace-webhook-server` deployment without removing the mutating and validating webhooks, you will not be able use `oc exec` commands to run commands in a container on the cluster. After you remove the webhooks you will be able to use the `oc exec` commands again.
+If you remove the `devworkspace-webhook-server` deployment without removing the mutating and validating webhooks, you will not be to able use `oc exec` commands to run commands in a container on the cluster. After you remove the webhooks you will be able to use the `oc exec` commands again.
```
Comment (Preeticp) [inline modules/odc-uninstalling-web-terminal.adoc:128]: you will not be able to
```suggestion
If you remove the `devworkspace-webhook-server` deployment without removing the mutating and validating webhooks, you will not be able to use `oc exec` commands to run commands in a container on the cluster. After you remove the webhooks you will be able to use the `oc exec` commands again.
```

**Comment by Preeticp** [comment]
@amisevsk two minor nits, please check. I am keeping the PR ready to merge with all the labels. Please ping me once you have fixed them, and I will merge the PR. Thank you!

File: modules/odc-uninstalling-web-terminal.adoc (hunk 4 above)
Comment (amisevsk) [reply modules/odc-uninstalling-web-terminal.adoc]: Hm, don't know how that got dropped -- sorry.

File: modules/odc-uninstalling-web-terminal.adoc (hunk 5 above)
Comment (amisevsk) [reply modules/odc-uninstalling-web-terminal.adoc]: :+1:

File: modules/odc-uninstalling-web-terminal.adoc (hunk 3 above)
Comment (amisevsk) [reply modules/odc-uninstalling-web-terminal.adoc]: Replaced with `remaining`

File: modules/odc-uninstalling-web-terminal.adoc (hunk 9 above)
Comment (amisevsk) [reply modules/odc-uninstalling-web-terminal.adoc:128]: :facepalm: Embarassing typo. Thanks

**Comment by amisevsk** [comment]
@Preeticp I think I've applied all current suggestions, thanks for the detailed review!

File: modules/odc-uninstalling-web-terminal.adoc (hunk 6 above)
Comment (amisevsk) [reply modules/odc-uninstalling-web-terminal.adoc]: :+1: thanks, I misunderstood the intention on `operator` vs `Operator` -- should be fixed now.

File: modules/odc-uninstalling-web-terminal.adoc (hunk 8 above)
[the remark this answers is not in the document]
Comment (amisevsk) [reply modules/odc-uninstalling-web-terminal.adoc]: I indeed did misunderstand, my apologies. I've updated all instances to `Operator` where it's used in a sentence. 

Regarding the last suggestion, I added `the DevWorkspace` and replaced `all the related` with `any related` -- "all the related" reads a little strangely to me.

**Comment by Preeticp** [comment]
Hi @amisevsk one last thing, please squash all your commits.
